### PR TITLE
changefeedccl: sqlsmith test

### PR DESCRIPTION
**Pull request by HonoreDB**
Informs https://github.com/cockroachdb/cockroach/issues/83591

Adds a test generating random predicates and
verifies that (except for a few known issues)
if a changefeed runs successfully it will output
the same rows as a regular query.

Release note: None

### CODE REVIEW

**Review by jayshrivastava — COMMENTED** [review]
Should we consider making this a roachtest since it's randomized and somewhat likely to fail (bc it's job is to look for edge cases)?

There's a lot of points of failure (ex. when first executing the cf statement, when executing the select statement, when asserting the payloads - and we repeat this 1000 times). Maybe it makes sense to add a boolean flag and set it to true if we see at least one error in any of those cases. We can also have an array of failures (errors + metadata / CREATE CHANGEFEED stmt / WHERE clause). At the end of the test, check the flag to see if there was at least one failure and log the array.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92345#-:-NIXlK1YBcc5dynRo0rs:b-34jj4m)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, and @stevendanna)

___
*[`pkg/internal/sqlsmith/relational.go` line 188 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/92345#-NIXmUPz9eRZDJTW9f3Y:-NIXmUQ098ThApD1X1Vd:b-rhuq1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a63e724040ba878e8db0e0ce351ed50711b5f3e/pkg/internal/sqlsmith/relational.go#L188)):*
> ```Go
> 		return nil, nil, false
> 	}
> ```

nit: extra space
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 951 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/92345#-NIXi54-D860LYElbweM:-NIXi54-D860LYElbweN:b-6yrus2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a63e724040ba878e8db0e0ce351ed50711b5f3e/pkg/ccl/changefeedccl/changefeed_test.go#L951)):*
> ```Go
> 			var expectedRowIDsStr string
> 			if err := rows.Scan(&expectedRowIDsStr); err != nil {
> 				t.Logf("Skipping query %s because error %s", query, err)
> ```

I feel like an error should not occur here. We might want to fail the test due to an unexpected error here.
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 964 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/92345#-NIXjdSV2_nOj2bcH7Pg:-NIXjdSV2_nOj2bcH7Ph:bk1dq1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a63e724040ba878e8db0e0ce351ed50711b5f3e/pkg/ccl/changefeedccl/changefeed_test.go#L964)):*
> ```Go
> 			defer closeFeedIgnoreError(t, seedFeed)
> 			if err != nil {
> 				t.Logf("Test tolerating create changefeed error: %s", err.Error())
> ```

If there's a changefeed expression that doesn't work here but works in a SELECT stmt, should we fail the test?
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 972 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/92345#-NIXk2fG3yKlidUxs2UZ:-NIXk2fG3yKlidUxs2U_:bkqr8ds) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a63e724040ba878e8db0e0ce351ed50711b5f3e/pkg/ccl/changefeedccl/changefeed_test.go#L972)):*
> ```Go
> 			}
> 			err = assertPayloadsBaseErr(context.Background(), seedFeed, assertedPayloads, false, false)
> 			if err != nil && (strings.Contains(err.Error(), "expected\n") || strings.Contains(err.Error(), "ordering guarantee")) {
> ```

Since we pass `false` into the last param of `assertPayloadsBaseErr`, is it still possible to have an "ordering guarantee" error?
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 975 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/92345#-NIXiv09-HPMFTb19xKI:-NIXiv09-HPMFTb19xKJ:bmzthz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a63e724040ba878e8db0e0ce351ed50711b5f3e/pkg/ccl/changefeedccl/changefeed_test.go#L975)):*
> ```Go
> 				t.Error(err)
> 			} else if err != nil {
> 				t.Logf("Test tolerating running changefeed error: %s", err.Error())
> ```

What kind of errors can happen here? I feel that transient errors won't happen in this test, so maybe fail in all cases here?

File: pkg/ccl/changefeedccl/changefeed_test.go [hunk 1]
```diff
@@ -65,8 +66,16 @@ import (
...
+	"github.com/cockroachdb/cockroach/pkg/sql/parser"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+<<<<<<< HEAD
+=======
+	"github.com/cockroachdb/cockroach/pkg/sql/randgen"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb"
+>>>>>>> 4a63e72404 (changefeedccl: sqlsmith test)
```
Comment (mgartner) [inline pkg/ccl/changefeedccl/changefeed_test.go:78]: You have some rebase remnants here.

File: pkg/sql/randgen/schema.go [hunk 2]
```diff
@@ -331,7 +331,10 @@ func PopulateTableWithRandData(
 				// them to the list of columns to insert data into.
 				continue
 			}
-			colTypes = append(colTypes, tree.MustBeStaticallyKnownType(col.Type.(*types.T)))
+			if _, ok := col.Type.(*types.T); !ok {
+				return 0, errors.Newf("No type for %v", col)
```
Comment (mgartner) [inline pkg/sql/randgen/schema.go:335]: Why was this change necessary?

**Review by mgartner — APPROVED** [review]
sqlsmith and randgen changes LGTM, just one question

File: pkg/sql/randgen/schema.go (hunk 2 above)
Comment (HonoreDB) [reply pkg/sql/randgen/schema.go:335]: iirc this was panicking on user-defined types without a helpful error message.

Comment (mgartner) [reply pkg/sql/randgen/schema.go:335]: Thanks for the explanation.

**Review by HonoreDB — COMMENTED** [review]
> Dismissed @mgartner from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92345)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jayshrivastava, @mgartner, @miretskiy, and @stevendanna)

___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 951 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/92345#-NIXi54-D860LYElbweM:-NKUETXk9bZflGhnji9H:b-5sze84) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a63e724040ba878e8db0e0ce351ed50711b5f3e/pkg/ccl/changefeedccl/changefeed_test.go#L951)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

I feel like an error should not occur here. We might want to fail the test due to an unexpected error here.
</blockquote></details>

sqlsmith, by design, often generates invalid SQL, so we need to tolerate errors here.
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 964 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/92345#-NIXjdSV2_nOj2bcH7Pg:-NKUEj0k6d-4KGM3kgMC:bqk4c88) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a63e724040ba878e8db0e0ce351ed50711b5f3e/pkg/ccl/changefeedccl/changefeed_test.go#L964)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

If there's a changefeed expression that doesn't work here but works in a SELECT stmt, should we fail the test?
</blockquote></details>

At the moment, no--for example, volatile functions should error on create.
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 972 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/92345#-NIXk2fG3yKlidUxs2UZ:-NKUEsGZBYxm4h7fDKLV:b-311uec) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a63e724040ba878e8db0e0ce351ed50711b5f3e/pkg/ccl/changefeedccl/changefeed_test.go#L972)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

Since we pass `false` into the last param of `assertPayloadsBaseErr`, is it still possible to have an "ordering guarantee" error?
</blockquote></details>

Shouldn't be, just being defensive.
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 975 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/92345#-NIXiv09-HPMFTb19xKI:-NKUExVKD9ihiGC1fp0I:b-9nabgl) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a63e724040ba878e8db0e0ce351ed50711b5f3e/pkg/ccl/changefeedccl/changefeed_test.go#L975)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

What kind of errors can happen here? I feel that transient errors won't happen in this test, so maybe fail in all cases here?
</blockquote></details>

There used to be a lot! But with recent changes I think I can remove this tolerance (and have).
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 78 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/92345#-NHVqYInMv6_H3GRzTqa-r4-78:-NKUER7B21bRzWUGdScf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b82996a252e5ea4f50160af4b65aff9f6557b1fc/pkg/ccl/changefeedccl/changefeed_test.go#L78)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

You have some rebase remnants here.
</blockquote></details>

Done.

**Comment by shermanCRL** [comment]
Waiting on anything to merge here?

**Review by jayshrivastava — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92345)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @mgartner, @miretskiy, and @stevendanna)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/92345#-NMerMbM3LrDIwTLvsqV:-NMerMbM3LrDIwTLvsqW:bpcibxr) (no related file):*
LGTM! Small comment below.

___
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 950 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/92345#-NMeqrPQ0lvtHKrOH6G5:-NMeqrPR26PDiIn0yRv1:bt1etux) ([raw file](https://github.com/cockroachdb/cockroach/blob/b19cfd201e105252a6718f154ce033d88f5dfa38/pkg/ccl/changefeedccl/changefeed_test.go#L950)):*
> ```Go
> 			err = assertPayloadsBaseErr(context.Background(), seedFeed, assertedPayloads, false, false)
> 			closeFeedIgnoreError(t, seedFeed)
> 			if err != nil && (strings.Contains(err.Error(), "expected\n") || strings.Contains(err.Error(), "ordering guarantee")) {
> ```

This if / if else block can probably be just 
```
if err != nil {
    t.Error(err)
}
```

**Comment by HonoreDB** [comment]
bors r=[mgartner,jayshrivastava]

**Comment by HonoreDB** [comment]
bors retry
